common/diagnostics/Metrics.java

```diff
@@ -6,203 +6,761 @@
+    public void incrementCurrentCount(@Nullable String databaseName, ConnectionPeakCounts.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        databaseLoad.get(databaseHash).incrementCurrent(kind);
+    }
+
+    public void decrementCurrentCount(@Nullable String databaseName, ConnectionPeakCounts.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        databaseLoad.get(databaseHash).decrementCurrent(kind);
+    }
+
+    public void registerError(@Nullable String databaseName, String errorCode) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        userErrors.get(databaseHash).register(errorCode);
     }
 
-    protected JsonObject asJSON() {
-        JsonObject metrics = new JsonObject();
-        metrics.add("system", system.asJSON());
-        metrics.add("requests", requests.asJSON());
-        metrics.add("DB usage", usage.asJSON());
-        metrics.add("user errors", userErrors.asJSON());
+    public void submitDatabaseDiagnostics(Set<DatabaseDiagnostics> databaseDiagnostics) {
+        Set<String> deletedDatabases = new HashSet<>(databaseLoad.keySet());
+
+        for (DatabaseDiagnostics diagnostics : databaseDiagnostics) {
+            String databaseHash = hashAndAddDatabaseIfAbsent(diagnostics.databaseName());
+            databaseLoad.get(databaseHash).setSchemaLoad(diagnostics.schemaLoad());
+            databaseLoad.get(databaseHash).setDataLoad(diagnostics.dataLoad());
+            updatePrimaryDatabases(databaseHash, diagnostics.isPrimaryServer());
+            deletedDatabases.remove(databaseHash);
+        }
+
+        deletedDatabases.forEach(databaseHash -> databaseLoad.get(databaseHash).setDatabaseDeleted());
+    }
+    
+    private static String formatDateTime(LocalDateTime dateTime) {
+        return dateTime.truncatedTo(ChronoUnit.SECONDS).toString();
+    }
+
+    protected JsonObject asJSON(boolean reporting) {
+        boolean isNoUsageDiagnostics = reporting && !base.getReportingEnabled();
+
+        JsonObject metrics = base.asJSON();
+        metrics.add("server", serverProperties.asJSON(isNoUsageDiagnostics));
+
+        if (isNoUsageDiagnostics) {
+            return metrics;
+        }
+
+        JsonArray load = new JsonArray();
+        databaseLoad.forEach((databaseHash, diagnostics) ->
+                diagnostics.asJSON(databaseHash, primaryDatabaseHashes.contains(databaseHash)).forEach(load::add));
+        metrics.add("load", load);
+
+        JsonArray actions = new JsonArray();
+        requests.forEach((databaseHash, diagnostics) -> diagnostics.asJSON(databaseHash).forEach(actions::add));
+        metrics.add("actions", actions);
+
+        JsonArray errors = new JsonArray();
+        userErrors.forEach((databaseHash, diagnostics) -> diagnostics.asJSON(databaseHash).forEach(errors::add));
+        metrics.add("errors", errors);
+
         return metrics;
     }
```

Comment:
You short-circuit here already, which means this can be split and let us avoid the opaque boolan flag:

```java
    protected JsonObject asFullJSON() {
        JsonObject metrics = base.asJSON();
        metrics.add("server", serverProperties.asFullJSON());
      
        // ...

        return metrics;
    }
    
    protected JsonObject asReportingJSON() {
        if (base.getReportingEnabled()) {
            return asFullJSON();
        } else {
            JsonObject metrics = base.asJSON();        
            metrics.add("server", serverProperties.asMinimalJSON());
            return metrics;
        }
    }
```
You only switch in a few places afaics. (Names are only a suggestion).

Reply:
Could be an option. Will do the final round once we understand the purpose of the monitoring page.

Reply:
Used it, thx.